packages/react/src/components/Checkbox/Checkbox.test.tsx

```diff
@@ -0,0 +1,185 @@
+
+type SetOptional<T, Keys extends keyof T> = Pick<Partial<T>, Keys> &
+  Omit<T, Keys>;
+type CheckboxProps = SetOptional<
+  React.ComponentProps<typeof Checkbox>,
+  'label' | 'id'
+>;
+
+const renderCheckbox = ({
+  label,
+  ...props
+}: CheckboxProps = {}): HTMLInputElement => {
+  render(<Checkbox id="checkbox-id" label={label || 'checkbox'} {...props} />);
+  // no individual assertions needed, as this should throw if the element does not exist
+  return screen.getByRole('checkbox', {
+    name: label as string
+  }) as HTMLInputElement;
+};
+
+test('should render unchecked checkbox', () => {
+  const input = renderCheckbox({ label: 'this is a checkbox' });
+  expect(input).not.toBeChecked();
+  expect(screen.queryByText('this is a checkbox')).toBeInTheDocument();
+});
+
+test('should render checked checkbox', () => {
+  const input = renderCheckbox({ checked: true });
+  expect(input).toBeChecked();
+});
+
+test('should render disabled unchecked checkbox', () => {
+  const input = renderCheckbox({ disabled: true });
+  expect(input).toBeDisabled();
+  expect(input).not.toBeChecked();
+});
+
+test('should render disabled checked checkbox', () => {
+  const input = renderCheckbox({ disabled: true, checked: true });
+  expect(input).toBeDisabled();
+  expect(input).toBeChecked();
+});
+
+test('should render error checkbox', () => {
+  const input = renderCheckbox({ error: 'you should check this checkbox' });
+  expect(input).toHaveAccessibleDescription('you should check this checkbox');
+  expect(
+    screen.queryByText('you should check this checkbox')
+  ).toBeInTheDocument();
+});
+
+test('should toggle checkbox correctly', async () => {
+  const user = userEvent.setup();
+  const input = renderCheckbox();
+  const checkboxIcon = input.parentElement!.querySelector(
+    '.Checkbox__overlay'
+  ) as HTMLElement;
+  expect(checkboxIcon).toHaveClass('Icon--checkbox-unchecked');
+  expect(checkboxIcon).not.toHaveClass('Icon--checkbox-checked');
+  expect(input).not.toBeChecked();
+
+  await user.click(checkboxIcon);
+  expect(checkboxIcon).not.toHaveClass('Icon--checkbox-unchecked');
+  expect(checkboxIcon).toHaveClass('Icon--checkbox-checked');
+});
+
+test('should handle focus correctly', () => {
+  const onFocus = spy();
+  const input = renderCheckbox({ onFocus });
+  const checkboxIcon = input.parentElement!.querySelector(
+    '.Checkbox__overlay'
+  ) as HTMLElement;
+  expect(checkboxIcon).not.toHaveClass('.Checkbox__overlay--focused');
+  expect(onFocus.notCalled).toBeTruthy();
+
+  input.focus();
+  expect(checkboxIcon).toHaveClass('Checkbox__overlay--focused');
+  expect(onFocus.calledOnce).toBeTruthy();
+});
+
+test('should handle blur correctly', () => {
+  const onBlur = spy();
+  const input = renderCheckbox({ onBlur, checked: true });
+  const checkboxIcon = input.parentElement!.querySelector(
+    '.Checkbox__overlay'
+  ) as HTMLElement;
+  expect(checkboxIcon).not.toHaveClass('.Checkbox__overlay--focused');
```

Comment:
Not a huge deal, but we can use `expect(checkboxIcon).toHaveFocus()` and `expect(checkboxIcon).not.toHaveFocus()` in addition to checking against classes here. It'd be a nice to make sure we're explicitly looking for the focus attribute.

Reply:
Yeah, totally fair. That's an easy change to make.